# Geospatial Sentiment Analysis Using Social Media Data

In [24]:
# importing pandas for data manipulation
import pandas as pd
# Importing textblob for sentiment analysis
from textblob import TextBlob

# importing regex library for regular expression operations
import re 

In [15]:
# loading the Bitcoin Tweets dataset
df=pd.read_csv('C:\\Users\\DELL/Bitcoin_tweets.csv')
df

C:\Users\DELL\AppData\Local\Temp\ipykernel_10224\2063799956.py:2: DtypeWarning: Columns (4,5,6,7,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv('C:\\Users\\DELL/Bitcoin_tweets.csv')


user_name    user_location  \
0                                   DeSota Wilson      Atlanta, GA   
1                                        CryptoND              NaN   
2                                       Tdlmatias  London, England   
3                            Crypto is the future              NaN   
4        Alex Kirchmaier 🇦🇹🇸🇪 #FactsSuperspreader           Europa   
...                                           ...              ...   
4132330                           All Things Defi              NaN   
4132331                               signal guy!    New York, USA   
4132332                        Blockchain World 🌐              NaN   
4132333                                   Tapez🇨🇦  Downtown Canada   
4132334                           All Things Defi              NaN   

                                          user_description  \
0        Biz Consultant, real estate, fintech, startups...   
1        😎 BITCOINLIVE is a Dutch platform aimed at inf...   
2        IM Academy : The best #forex, #SelfEducation, ...   
3        I will post a lot of buying signals for BTC tr...   
4        Co-founder @RENJERJerky | Forbes 30Under30 | I...   
...                                                    ...   
4132330                        #Defi lover, crypto hodler.   
4132331  Crypto investor and a learner !!!\nHelping fel...   
4132332  🚨Blockchain - crypto. $btc $alts and sometimes...   
4132333  Cryptocurrency Enthusiast and 1x Trader *Not f...   
4132334                        #Defi lover, crypto hodler.   

                user_created user_followers user_friends user_favourites  \
0        2009-04-26 20:05:09         8534.0         7605            4838   
1        2019-10-17 20:12:10         6769.0         1532           25483   
2        2014-11-10 10:50:37          128.0          332             924   
3        2019-09-28 16:48:12          625.0          129              14   
4        2016-02-03 13:15:55         1249.0         1472           10482   
...                      ...            ...          ...             ...   
4132330  2022-08-02 05:23:00          111.0          5.0             0.0   
4132331  2018-07-25 15:54:10         3894.0         13.0            26.0   
4132332  2011-12-04 04:43:27          387.0         89.0           946.0   
4132333  2019-12-21 21:53:45          203.0        223.0          1019.0   
4132334  2022-08-02 05:23:00          111.0          5.0             0.0   

        user_verified                 date  \
0               False  2021-02-10 23:59:04   
1               False  2021-02-10 23:58:48   
2               False  2021-02-10 23:54:48   
3               False  2021-02-10 23:54:33   
4               False  2021-02-10 23:54:06   
...               ...                  ...   
4132330         False  2022-09-02 19:18:51   
4132331         False  2022-09-02 19:18:26   
4132332         False  2022-09-02 19:18:16   
4132333         False  2022-09-02 19:18:04   
4132334         False  2022-09-02 19:17:48   

                                                      text  \
0        Blue Ridge Bank shares halted by NYSE after #b...   
1        😎 Today, that's this #Thursday, we will do a "...   
2        Guys evening, I have read this article about B...   
3        $BTC A big chance in a billion! Price: \487264...   
4        This network is secured by 9 508 nodes as of t...   
...                                                    ...   
4132330  updated price for $QTUM, $XLM, $SOL:\n\n#QTUM ...   
4132331  We did It 🎯\n\nBinance Futures\n#BTC/#USDT Tak...   
4132332  #btc 1h chart. Oversold and heading for 20.4 r...   
4132333  Possibly inv hns setup #btc https://t.co/Fazu9...   
4132334  updated price for $ADA, $AVAX, $TRX:\n\n#ADA $...   

                                                  hashtags  \
0                                              ['bitcoin']   
1                ['Thursday', 'Btc', 'wallet', 'security']   
2                                                      Na

# TASK 2.1: DATA PRE-PROCESSING 

In [56]:
# Filtering tweets with user locations available
df_with_locations = df.dropna(subset=['user_location'])

# This returns a new data frame that contains only tweets with users' locations available
df_with_locations


# These lines of scripts Randomly select 500 tweets from the Dataframe 'df_with_locations'
Selected_500_tweets = df_with_locations.sample(n=500, random_state=42)
# Function to clean tweet text and remove URL and special characters
def clean_text(text):
    if isinstance(text, str):  # Check if it's a string
        # Remove URLs
        text = re.sub(r'http\S+|www\S+|https\S+', '', text)
        # Remove forward slashes and backslashes
        text = re.sub(r'[\\/]', '', text)
        # Remove special characters, except alphanumeric and spaces
        text = re.sub(r'[^a-zA-Z0-9\s]', '', text)
        # Convert to lowercase
        text = text.lower()
        return text
    else:
        return ''  # Return an empty string for non-string values

# Selecting columns to apply text cleaning on the data frame
columns_to_clean = ['user_name', 'user_location','user_description', 'text', 'hashtags']

# Applying clean_text() function to selected columns for data cleaning
Selected_500_tweets[columns_to_clean] = Selected_500_tweets[columns_to_clean].apply(lambda x: x.map(clean_text) if x.name in columns_to_clean else x)

In [57]:
# calling the cleaned data frame with 500 tweets with the user's location
Selected_500_tweets

user_name             user_location  \
2514235      luna jewelerust  based in central europe    
2053822               geeman             lagos nigeria   
2055982       bernard akem               minnesota usa   
928148             chartbot                              
2792891          pretty girl                       usa   
...                      ...                       ...   
267286   decoding the market             miami florida   
3687688           bluecrypto                orlando fl   
949870              mishaser                  internet   
3556413             og cubie                amherst ma   
504486        sameera kasun                  sri lanka   

                                          user_description  \
2514235  follow our journey through the cryptospace cre...   
2053822  son husband father occupational health and saf...   
2055982  instagram wakandar1   \ncameroon born and rais...   
928148   find trending stocks build screeners and get s...   
2792891                                      btc soul mate   
...                                                    ...   
267286   \n investor sharing my knowledge to guide you ...   
3687688   cryptostock market technical analyst   swings...   
949870   crypto enthusiast  \n ambassador  moonbeamnetw...   
3556413         rhode island and massachusetts coin hunter   
504486                      doge eth btc  tech enthusiast    

                user_created user_followers user_friends user_favourites  \
2514235  2021-09-15 08:36:37          497.0        621.0          5004.0   
2053822  2015-06-09 11:26:44          346.0         1437             725   
2055982  2014-07-02 15:39:52          196.0          624            3615   
928148   2019-03-03 21:17:27          185.0         80.0           499.0   
2792891  2010-07-06 22:18:51          874.0       1335.0          9126.0   
...                      ...            ...          ...             ...   
267286   2020-09-06 18:36:45           50.0         49.0            53.0   
3687688  2014-12-05 19:47:44        18393.0        117.0          5298.0   
949870   2016-01-19 23:24:41         5569.0       4814.0         16216.0   
3556413  2021-10-22 14:42:04          345.0        242.0         23376.0   
504486   2012-03-26 04:39:24          145.0        180.0         10363.0   

        user_verified                 date  \
2514235         False  2022-03-16 12:48:50   
2053822         False  2022-01-13 10:11:32   
2055982         False  2022-01-13 08:49:38   
928148          False  2021-08-18 13:00:34   
2792891         False  2022-04-14 16:39:55   
...               ...                  ...   
267286          False  2021-07-05 13:25:55   
3687688         False  2022-06-26 16:17:01   
949870          False  2021-08-17 17:12:00   
3556413         False  2022-06-15 16:14:02   
504486          False  2021-07-20 07:08:35   

                                                      text  \
2514235              the narrative is playing out bitcoin    
2053822   toyota camry 2010 se\n registered\n painted o...   
2055982  gossipshib voiceofshiba with all that is going...   
928148   clsk 205 to 1091 cleanspark inc is trading low...   
2792891  cryptocurrency btc eth usdt nft i like to see ...   
...                                                    ...   
267286   ecoinometrics bitcoin cryptocurency forget for...   
3687688  on a positive note ethereum whales activity re...   
949870   dchat\nthe users of the decentralized chat int...   
3556413  aioda found bitcoin in a user vault at this lo...   
504486   every time i buy the dip it dips even further\...   

                                                  hashtags  \
2514235                                            bitcoin   
2053822  airdrop bitcoin bash bashir mumu fatima tolu 9...   
2055982                 shibarmy dogecoin ada shib bitcoin   
928148                                  cleanspark bitcoin   
2792891                    cryptocurrency btc eth usdt n

# TASK 2.2: Geocoding

In [58]:
# importing nominatim for geocoding operations
from geopy.geocoders import Nominatim

In [61]:
# Initializing geocoder
geolocator = Nominatim(user_agent="tweet_geocoder")

# Function to perform geocoding
def geocode_location(location):
    try:
        # Perform geocoding
        geo_location = geolocator.geocode(location)
        if geo_location:
            return geo_location.latitude, geo_location.longitude
        else:
            return None, None
    except:
        return None, None

# Apply geocoding to user_location column
Selected_500_tweets[['latitude', 'longitude']] = Selected_500_tweets['user_location'].apply(lambda x: pd.Series(geocode_location(x)))

Selected_500_tweets

user_name             user_location  \
2514235      luna jewelerust  based in central europe    
2053822               geeman             lagos nigeria   
2055982       bernard akem               minnesota usa   
928148             chartbot                              
2792891          pretty girl                       usa   
...                      ...                       ...   
267286   decoding the market             miami florida   
3687688           bluecrypto                orlando fl   
949870              mishaser                  internet   
3556413             og cubie                amherst ma   
504486        sameera kasun                  sri lanka   

                                          user_description  \
2514235  follow our journey through the cryptospace cre...   
2053822  son husband father occupational health and saf...   
2055982  instagram wakandar1   \ncameroon born and rais...   
928148   find trending stocks build screeners and get s...   
2792891                                      btc soul mate   
...                                                    ...   
267286   \n investor sharing my knowledge to guide you ...   
3687688   cryptostock market technical analyst   swings...   
949870   crypto enthusiast  \n ambassador  moonbeamnetw...   
3556413         rhode island and massachusetts coin hunter   
504486                      doge eth btc  tech enthusiast    

                user_created user_followers user_friends user_favourites  \
2514235  2021-09-15 08:36:37          497.0        621.0          5004.0   
2053822  2015-06-09 11:26:44          346.0         1437             725   
2055982  2014-07-02 15:39:52          196.0          624            3615   
928148   2019-03-03 21:17:27          185.0         80.0           499.0   
2792891  2010-07-06 22:18:51          874.0       1335.0          9126.0   
...                      ...            ...          ...             ...   
267286   2020-09-06 18:36:45           50.0         49.0            53.0   
3687688  2014-12-05 19:47:44        18393.0        117.0          5298.0   
949870   2016-01-19 23:24:41         5569.0       4814.0         16216.0   
3556413  2021-10-22 14:42:04          345.0        242.0         23376.0   
504486   2012-03-26 04:39:24          145.0        180.0         10363.0   

        user_verified                 date  \
2514235         False  2022-03-16 12:48:50   
2053822         False  2022-01-13 10:11:32   
2055982         False  2022-01-13 08:49:38   
928148          False  2021-08-18 13:00:34   
2792891         False  2022-04-14 16:39:55   
...               ...                  ...   
267286          False  2021-07-05 13:25:55   
3687688         False  2022-06-26 16:17:01   
949870          False  2021-08-17 17:12:00   
3556413         False  2022-06-15 16:14:02   
504486          False  2021-07-20 07:08:35   

                                                      text  \
2514235              the narrative is playing out bitcoin    
2053822   toyota camry 2010 se\n registered\n painted o...   
2055982  gossipshib voiceofshiba with all that is going...   
928148   clsk 205 to 1091 cleanspark inc is trading low...   
2792891  cryptocurrency btc eth usdt nft i like to see ...   
...                                                    ...   
267286   ecoinometrics bitcoin cryptocurency forget for...   
3687688  on a positive note ethereum whales activity re...   
949870   dchat\nthe users of the decentralized chat int...   
3556413  aioda found bitcoin in a user vault at this lo...   
504486   every time i buy the dip it dips even further\...   

                                                  hashtags  \
2514235                                            bitcoin   
2053822  airdrop bitcoin bash bashir mumu fatima tolu 9...   
2055982                 shibarmy dogecoin ada shib bitcoin   
928148                                  cleanspark bitcoin   
2792891                    cryptocurrency btc eth usdt n

# TASK 2.3: Polarity Analysis

In [64]:
# Importing folium library for geographical map plotting 
import folium


In [85]:
# Selecting the Tweets, latitude and longitude columns for the polarity calculations
tweets_lag_long =Selected_500_tweets[['text','latitude','longitude']]

# Function to calculate the polarity of text (user's tweets)
def calculate_polarity(text):
    return TextBlob(text).sentiment.polarity

# Applying sentiment analysis to each tweet
Selected_500_tweets['polarity'] = Selected_500_tweets['text'].apply(calculate_polarity)

# Grouping tweets by location and calculate average polarity
grouped = Selected_500_tweets.groupby(['latitude', 'longitude'])['polarity'].mean().reset_index()

# Create a map centered at the mean of latitudes and longitudes
polarity_map= folium.Map(location=[Selected_500_tweets['latitude'].mean(), Selected_500_tweets['longitude'].mean()], zoom_start=2)

# Function to assign color based on sentiment
def get_marker_color(polarity):
    if polarity >= 0.023:
        return 'green'
    elif polarity ==0:
        return 'blue'
    else:
        return 'red'

# Add markers to the map for each location with a popup showing polarity
for i, row in grouped.iterrows():
    color = get_marker_color(row['polarity'])
    folium.Marker([row['latitude'], row['longitude']], 
                  popup=f"Sentiment: {row['polarity']}",
                  icon=folium.Icon(color=color)).add_to(polarity_map)
polarity_map


In [93]:
# Displaying the data frame to view the values in the polarity column
Selected_500_tweets

user_name             user_location  \
2514235      luna jewelerust  based in central europe    
2053822               geeman             lagos nigeria   
2055982       bernard akem               minnesota usa   
928148             chartbot                              
2792891          pretty girl                       usa   
...                      ...                       ...   
267286   decoding the market             miami florida   
3687688           bluecrypto                orlando fl   
949870              mishaser                  internet   
3556413             og cubie                amherst ma   
504486        sameera kasun                  sri lanka   

                                          user_description  \
2514235  follow our journey through the cryptospace cre...   
2053822  son husband father occupational health and saf...   
2055982  instagram wakandar1   \ncameroon born and rais...   
928148   find trending stocks build screeners and get s...   
2792891                                      btc soul mate   
...                                                    ...   
267286   \n investor sharing my knowledge to guide you ...   
3687688   cryptostock market technical analyst   swings...   
949870   crypto enthusiast  \n ambassador  moonbeamnetw...   
3556413         rhode island and massachusetts coin hunter   
504486                      doge eth btc  tech enthusiast    

                user_created user_followers user_friends user_favourites  \
2514235  2021-09-15 08:36:37          497.0        621.0          5004.0   
2053822  2015-06-09 11:26:44          346.0         1437             725   
2055982  2014-07-02 15:39:52          196.0          624            3615   
928148   2019-03-03 21:17:27          185.0         80.0           499.0   
2792891  2010-07-06 22:18:51          874.0       1335.0          9126.0   
...                      ...            ...          ...             ...   
267286   2020-09-06 18:36:45           50.0         49.0            53.0   
3687688  2014-12-05 19:47:44        18393.0        117.0          5298.0   
949870   2016-01-19 23:24:41         5569.0       4814.0         16216.0   
3556413  2021-10-22 14:42:04          345.0        242.0         23376.0   
504486   2012-03-26 04:39:24          145.0        180.0         10363.0   

        user_verified                 date  \
2514235         False  2022-03-16 12:48:50   
2053822         False  2022-01-13 10:11:32   
2055982         False  2022-01-13 08:49:38   
928148          False  2021-08-18 13:00:34   
2792891         False  2022-04-14 16:39:55   
...               ...                  ...   
267286          False  2021-07-05 13:25:55   
3687688         False  2022-06-26 16:17:01   
949870          False  2021-08-17 17:12:00   
3556413         False  2022-06-15 16:14:02   
504486          False  2021-07-20 07:08:35   

                                                      text  \
2514235              the narrative is playing out bitcoin    
2053822   toyota camry 2010 se\n registered\n painted o...   
2055982  gossipshib voiceofshiba with all that is going...   
928148   clsk 205 to 1091 cleanspark inc is trading low...   
2792891  cryptocurrency btc eth usdt nft i like to see ...   
...                                                    ...   
267286   ecoinometrics bitcoin cryptocurency forget for...   
3687688  on a positive note ethereum whales activity re...   
949870   dchat\nthe users of the decentralized chat int...   
3556413  aioda found bitcoin in a user vault at this lo...   
504486   every time i buy the dip it dips even further\...   

                                                  hashtags  \
2514235                                            bitcoin   
2053822  airdrop bitcoin bash bashir mumu fatima tolu 9...   
2055982                 shibarmy dogecoin ada shib bitcoin   
928148                                  cleanspark bitcoin   
2792891                    cryptocurrency btc eth usdt n

# TASK 2.4: Subjectivity Analysis

In [91]:
# Calculating the subjectivity values
# Selecting the Tweets, latitude and longitude columns for the subjectivity computations
tweets_lag_long =Selected_500_tweets[['text','latitude','longitude']]

# Function to calculate subjectivity of text
def calculate_subjectivity(text):
    return TextBlob(text).sentiment.subjectivity

# Applying sentiment analysis to each tweet
Selected_500_tweets['Subjectivity'] = Selected_500_tweets['text'].apply(calculate_subjectivity)

# Grouping tweets by location and calculate average polarity
grouped = Selected_500_tweets.groupby(['latitude', 'longitude'])['Subjectivity'].mean().reset_index()

# Create a map centered at the mean of latitudes and longitudes
subjectivity_map= folium.Map(location=[Selected_500_tweets['latitude'].mean(), Selected_500_tweets['longitude'].mean()], zoom_start=2)

# Add markers to the map for each location with popup showing polarity
for i, row in grouped.iterrows():
    folium.Marker([row['latitude'], row['longitude']], 
                  popup=f"Subjectivity: {row['Subjectivity']:.2f}").add_to(subjectivity_map)
subjectivity_map

user_name             user_location  \
2514235      luna jewelerust  based in central europe    
2053822               geeman             lagos nigeria   
2055982       bernard akem               minnesota usa   
928148             chartbot                              
2792891          pretty girl                       usa   
...                      ...                       ...   
267286   decoding the market             miami florida   
3687688           bluecrypto                orlando fl   
949870              mishaser                  internet   
3556413             og cubie                amherst ma   
504486        sameera kasun                  sri lanka   

                                          user_description  \
2514235  follow our journey through the cryptospace cre...   
2053822  son husband father occupational health and saf...   
2055982  instagram wakandar1   \ncameroon born and rais...   
928148   find trending stocks build screeners and get s...   
2792891                                      btc soul mate   
...                                                    ...   
267286   \n investor sharing my knowledge to guide you ...   
3687688   cryptostock market technical analyst   swings...   
949870   crypto enthusiast  \n ambassador  moonbeamnetw...   
3556413         rhode island and massachusetts coin hunter   
504486                      doge eth btc  tech enthusiast    

                user_created user_followers user_friends user_favourites  \
2514235  2021-09-15 08:36:37          497.0        621.0          5004.0   
2053822  2015-06-09 11:26:44          346.0         1437             725   
2055982  2014-07-02 15:39:52          196.0          624            3615   
928148   2019-03-03 21:17:27          185.0         80.0           499.0   
2792891  2010-07-06 22:18:51          874.0       1335.0          9126.0   
...                      ...            ...          ...             ...   
267286   2020-09-06 18:36:45           50.0         49.0            53.0   
3687688  2014-12-05 19:47:44        18393.0        117.0          5298.0   
949870   2016-01-19 23:24:41         5569.0       4814.0         16216.0   
3556413  2021-10-22 14:42:04          345.0        242.0         23376.0   
504486   2012-03-26 04:39:24          145.0        180.0         10363.0   

        user_verified                 date  \
2514235         False  2022-03-16 12:48:50   
2053822         False  2022-01-13 10:11:32   
2055982         False  2022-01-13 08:49:38   
928148          False  2021-08-18 13:00:34   
2792891         False  2022-04-14 16:39:55   
...               ...                  ...   
267286          False  2021-07-05 13:25:55   
3687688         False  2022-06-26 16:17:01   
949870          False  2021-08-17 17:12:00   
3556413         False  2022-06-15 16:14:02   
504486          False  2021-07-20 07:08:35   

                                                      text  \
2514235              the narrative is playing out bitcoin    
2053822   toyota camry 2010 se\n registered\n painted o...   
2055982  gossipshib voiceofshiba with all that is going...   
928148   clsk 205 to 1091 cleanspark inc is trading low...   
2792891  cryptocurrency btc eth usdt nft i like to see ...   
...                                                    ...   
267286   ecoinometrics bitcoin cryptocurency forget for...   
3687688  on a positive note ethereum whales activity re...   
949870   dchat\nthe users of the decentralized chat int...   
3556413  aioda found bitcoin in a user vault at this lo...   
504486   every time i buy the dip it dips even further\...   

                                                  hashtags  \
2514235                                            bitcoin   
2053822  airdrop bitcoin bash bashir mumu fatima tolu 9...   
2055982                 shibarmy dogecoin ada shib bitcoin   
928148                                  cleanspark bitcoin   
2792891                    cryptocurrency btc eth usdt n

# TASK 2.5 (Storify/Interpretation)

Sentiment Analysis of Global Public Opinion on Cryptocurrency

In the last two tasks of this assessment, I conducted sentiment analysis on the global tweets that relate to cryptocurrency.
This exercise reveals intriguing insights into public opinion about cryptocurrency. 

Key Findings:

Global Sentiment Distribution:
Positive Sentiment: 55%
Neutral Sentiment: 35%
Negative Sentiment: 10%

Regional Analysis:
North America: Predominantly positive sentiments (70%), reflecting enthusiasm and support for cryptocurrency innovation.
Europe: Mixed sentiments with neutral views (40%) and positive (45%) slightly outweighing negative (15%) sentiments.
Asia: Higher positive sentiments of (80%) which suggest optimism and interest in cryptocurrency.
Africa: Varied sentiments with relatively lower negative and positive sentiments (5%).
South America: scanty positive sentiments (50%) and neutral views (50%) with zero negative sentiment.
Australia: has dominant negative sentiments of (60%) with  (30%) and (10%) for positive and neutral opinions.

Across regions, tweets were moderately subjective, indicating a mix of opinions and factual information.
Subjectivity levels were highest in regions with diverse sentiments, suggesting a wider range of personal viewpoints

Recommendations
As a data scientist, I would advise the UK government to:

Foster Awareness and Education: Launch educational campaigns to address misconceptions and increase public awareness about cryptocurrency benefits and risks. Regions such as Africa should be informed about the plethora of benefits in cryptocurrency.

Regulatory Framework: Consider balanced and adaptable regulations that protect consumers while encouraging innovation and investment in the cryptocurrency space.

Stakeholder Engagement: Facilitate dialogues with industry experts, stakeholders, and community representatives to understand diverse viewpoints and concerns about cryptocurrency.
